### PR TITLE
Initial code completion for SOQL

### PR DESCRIPTION
**Pull request by dobladez**
### What does this PR do?

Implements code-completion for SOQL.
* It supports the most common keywords, completion of `FROM` clause with SObject names and completion of `SELECT` fields with SObject fields.
* Basic support for nested SELECTs

Also:
 * Extract interaction with other Salesforce extensions into separate `sfdx.ts`
 * On SOQL builder and editor only propose "queryable" objects (fixes forcedotcom/soql-tooling#83)

Depends on https://github.com/forcedotcom/soql-tooling/pull/99

### What issues does this PR fix or reference?
@W-8431809@, @W-8432064@

### Functionality After
soql_completion_pr

### CODE REVIEW

File: packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts
Comment (dehru) [inline packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts:173]: Is there a reason you are only including 'NOT_FOUND'?

File: packages/salesforcedx-vscode-soql/src/sfdx/index.ts
Comment (dehru) [inline packages/salesforcedx-vscode-soql/src/sfdx/index.ts:23]: Why is this necessary?  Wouldn't this be fatal?

File: packages/salesforcedx-vscode-soql/src/sfdx/index.ts
Comment (dehru) [inline packages/salesforcedx-vscode-soql/src/sfdx/index.ts:37]: Thank you for this.

File: packages/salesforcedx-vscode-soql/src/sfdx/index.ts
Comment (dehru) [inline packages/salesforcedx-vscode-soql/src/sfdx/index.ts:44]: Oh, this is helpful.  What kinds of objects did this filter out?

File: packages/salesforcedx-vscode-soql/src/sfdx/index.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/sfdx/index.ts:23]: Good point. I simply moved the existing code without giving it much thought.

I'll add an assertion/safety check. The Extension should fail at activation time if that's `undefined`.

File: packages/salesforcedx-vscode-soql/src/sfdx/index.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/sfdx/index.ts:44]: Some SObjects cannot be queried for.  See: https://github.com/forcedotcom/soql-tooling/issues/83

File: packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts:173]: Hmm... for `SOBJECTS_REQUEST ` it probably shouldn't be done.

For the case of `SOBJECT_METADATA_REQUEST` however, what happens is that the query text might refer to a non-existent SObject. IMHO, we shouldn't show an error toast here... it's an error we should expect (say, while the user is typing a FROM clause on the editor) and should handle explicitly, for example: disabling the Fields selection dropdown, and may be show some explanation text (within the SOQL UI) that an object with that name could not be found).

Comment (dehru) [reply packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts:173]: Gotcha.  I won't send 'SOBJECT_METADATA_REQUEST' to telemetry either or print it in the channel.  I added a story to handle the case in the UI.

And if they do run the query, they will see the error that the field doesn't exist anyway.  The error information coming back from run query is informative.

File: packages/salesforcedx-vscode-soql/src/sfdx/index.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/sfdx/index.ts]: done

**Review by dehru — APPROVED** [review]

**Comment by jgellin-sf** [comment]
This PR is QA complete. We are ready to merge once we get a codeowner approval. @forcedotcom/pdt

File: .vscode/launch.json
Comment (lcampos) [inline .vscode/launch.json:619]:
```suggestion
      "preLaunchTask": "Compile",
```

File: packages/salesforcedx-vscode-soql/src/client/client.ts
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/src/client/client.ts:155]: Let's use a more meaningful name instead of just calling this `m`

File: packages/salesforcedx-vscode-soql/src/sfdx.ts
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/src/sfdx.ts:18]: We shouldn't be using this pattern anymore, replace it with the modules on `salesforcedx-utils-vscode` e.g. https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/packages/salesforcedx-utils-vscode/src/context/workspaceContextUtil.ts

File: packages/salesforcedx-vscode-soql/src/sfdx.ts
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/src/sfdx.ts:8]: What is the advantage of having all the services in one file like this ?

File: packages/salesforcedx-vscode-soql/test/vscode-integration/testUtilities.ts
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/test/vscode-integration/testUtilities.ts:27]: Same as above, let's stop using this pattern for depending on the core extension.

File: packages/salesforcedx-vscode-soql/src/client/client.ts
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/src/client/client.ts:63]: Can you provide an example on when this code is needed ? Which query and completions would I have to write to use this ?

**Review by lcampos — CHANGES_REQUESTED** [review]
Update the code to use salesforcedx-utils-vscode instead of depending on the core extension

**Comment by jgellin-sf** [comment]
@lcampos I have merged some preparatory PR's refactoring core/utils stuff so that we can finally sever the core dependency from the soql extension. Since this PR refactors the client, I was waiting for this PR to merge before finally removing the core dependency from this extension.

File: packages/salesforcedx-vscode-soql/src/sfdx.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/sfdx.ts:8]: My idea was to minimize and put together all integration points from the SOQL extensions to the other Salesforce extensions. To reduce code duplication, simplify testing and reduce coupling from the SOQL extension to the other salesforce extensions.

For instance, the logic to load the core extension (which we have to change), is now on a single place instead of four.

File: packages/salesforcedx-vscode-soql/src/sfdx.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/sfdx.ts:18]: Yes.  @jgellin-sf  has this covered on another PR on top of this one

File: packages/salesforcedx-vscode-soql/src/client/client.ts
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/client/client.ts:63]: Not sure I understand the question... but let me try to give some context.

The connection to the Salesforce Org (to retrieve SObjects metadata, needed for completion) is already handled by the Extensions, where the LSP client runs.

So, we thought of some options:

1. Don't use LSP at all and implement completion inside the VSCode extension. Problem gone.
2. The LSP server could use its own separate connection. This would add some complexity to pass on the connection info/credentials from the extension to the server, making sure the right Org is hit (in sync with the extension) and prevent us from sharing caching.
3. The LSP server asks the client for metadata when needed. This would make each code-completion requests very chatty:
image
4. The LSP server returns special completion items that the client must interpret and replace ("expand") with sobjects metadata:
image

So, we went with option 4, which allows us to continue using LSP while reusing the connections/caching of the extensions. What you see on this PR is the client-side of that.

More info here: https://salesforce.quip.com/JxYxAoa17YxO#NMUACAGfgxx

I hope my explanation makes sense. I would be happy to have a live discussion and go over more details. We already have improvements and refactorings coming on another PR on top of this one. Thanks!

Comment (lcampos) [reply packages/salesforcedx-vscode-soql/src/client/client.ts:63]: @dobladez could you give me access to that document ?

We should talk about how you are approaching the LSP functionality. Right now you are asking the client to massage/process the data returned by it. This might be ok for the SOQL extension but will make the SOQL LSP difficult to use in other scenarios/tools. At the moment, we know of at least two other tools outside of the SOQL extension where we want to use the language server and I don't believe the current approach is the correct one.

@praksb @jgellin-sf @dobladez lets talk about this once we're back from the holidays.

Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/client/client.ts:63]: Sure. Quip tells me you already have access to the quip doc....  anyway, the document doesn't say much more regarding connection-handling.

File: packages/salesforcedx-vscode-soql/src/sfdx.ts
Comment (jgellin-sf) [reply packages/salesforcedx-vscode-soql/src/sfdx.ts:18]: @lcampos FYI here is the PR for reference: https://github.com/forcedotcom/salesforcedx-vscode/pull/2837

File: packages/salesforcedx-vscode-soql/test/vscode-integration/testUtilities.ts
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/test/vscode-integration/testUtilities.ts:176]: nit: remove commented out code

**Review by lcampos — APPROVED** [review]

File: packages/salesforcedx-vscode-soql/src/client/client.ts
Comment (lcampos) [reply packages/salesforcedx-vscode-soql/src/client/client.ts:63]: circling back, we met and decided we'll change this in a future PR since it requires changes to the language server.

**Pull request merged**
